### PR TITLE
Remove, postpone, complete Task, restructuring.

### PR DESCRIPTION
**Pull request by RodionShyshkin**
Memory model classes were restructured, remove, postpone and completeTask methods were added. 

Documentation and tests were updated.

### CODE REVIEW

File: src/API/TaskService.cpp [hunk 1]
```diff
@@ -4,74 +4,62 @@
...
- * @param [none]
- *
- * @return vector of TaskDTO instances which forms all tasks for current day (local date).
- */
-
 std::vector<TaskDTO> TaskService::getTasksForToday() {
-  return task_service_storage_.GetTasksForToday();
+  auto todayTasks = task_service_storage_.GetTaskView().GetTodayTasks();
+  std::vector<TaskDTO> searchResult;
+  for(auto task : todayTasks) {
```
Comment (ValeriiSudakov) [inline src/API/TaskService.cpp]: it's better to use const auto&

File: src/MemoryModel/Storage/FullStorage.cpp [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  return task_storage_;
+}
+
+OperationResult FullStorage::AddTask(const Task &task) {
+  auto newid = generate_id_.GenerateID();
+  if(!newid.has_value()) return OperationResult::create(ErrorCode::MEMORY_LIMIT);
+  auto newtask = TaskEntity::createTask(task, newid.value());
+  auto task_ptr = std::make_shared<TaskEntity>(newtask);
+
+  if(!task_storage_.PushTask(task_ptr)) return OperationResult::create(ErrorCode::WRONG_TASK_ID);
```
Comment (ValeriiSudakov) [inline src/MemoryModel/Storage/FullStorage.cpp]: PushTask.. you should to rename

File: src/MemoryModel/Storage/TaskView.cpp [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+bool findByID(const TaskID& id, std::multimap<T, std::pair<TaskID, std::weak_ptr<TaskEntity>>> map) {
+  for(auto [first, second] : map) {
+    if(second.first == id) return true;
+  }
+  return false;
+}
+
+template <typename T>
+bool removeFromMap(const TaskID& id, std::multimap<T, std::pair<TaskID, std::weak_ptr<TaskEntity>>> map) {
+  for(auto it = map.begin(); it != map.end(); ++it) {
```
Comment (ValeriiSudakov) [inline src/MemoryModel/Storage/TaskView.cpp]: why dont u use const auto& : map ?

File: src/MemoryModel/Storage/TaskView.cpp [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+  return true;
+}
+
+bool TaskView::RemoveTask(const TaskID& id) {
+  bool flag = false;
+
+  if(!removeFromMap(id, priority_sorted_)) return false;
+  if(!removeFromMap(id, name_sorted_)) return false;
+  if(!removeFromMap(id, label_sorted_)) return false;
+  if(!removeFromMap(id, date_sorted_)) return false;
```
Comment (ValeriiSudakov) [inline src/MemoryModel/Storage/TaskView.cpp]: If 1 of remove returned false - other will not be removed...

File: src/MemoryModel/Storage/TaskView.cpp [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+  return searchResult;
+}
+
+std::vector<TaskEntity> TaskView::GetTasksByPriority(const Priority &priority) {
+  auto first = priority_sorted_.find(priority);
+  std::vector<TaskEntity> searchResult;
+
+  if(first == priority_sorted_.end()) return searchResult;
+  else {
+    for(auto it = first; it != priority_sorted_.end(); ++it) {
```
Comment (ValeriiSudakov) [inline src/MemoryModel/Storage/TaskView.cpp]: Why do u use different iterators in same loops?

File: src/MemoryModel/Task/TaskEntity.cpp [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+TaskID TaskEntity::GetID() const { return user_id_; }
+std::string TaskEntity::GetName() const { return task_.GetName(); }
+std::string TaskEntity::GetLabel() const { return task_.GetLabel(); }
+Priority TaskEntity::GetPriority() const { return task_.GetPriority(); }
+Date TaskEntity::GetDueTime() const { return task_.GetDate(); }
+bool TaskEntity::GetStatus() const { return status_; }
+Task TaskEntity::GetTask() const { return task_; }
+std::map<TaskID, std::shared_ptr<TaskEntity>> TaskEntity::GetSubtasks() const { return subtasks_; }
+TaskID TaskEntity::GetParentID() const { return parent_id_; }
```
Comment (ValeriiSudakov) [inline src/MemoryModel/Task/TaskEntity.cpp]: Add spaces between methods..............

File: src/MemoryModel/Task/TaskEntity.h [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+  bool                                                    checkParent();
+
+ private:
+  TaskEntity(const Task &task, const TaskID &id, const TaskID& parent);
+
+ private:
+  TaskID                                                  user_id_;
+  bool                                                    status_;
+  Task                                                    task_;
+  std::map<TaskID, std::shared_ptr<TaskEntity>>           subtasks_;
```
Comment (ValeriiSudakov) [inline src/MemoryModel/Task/TaskEntity.h]: R u sure, that shared ptr is necessary here? mb weak ptr?

**Review by ValeriiSudakov — CHANGES_REQUESTED** [review]

**Review by IliaIlyin — APPROVED** [review]

**Review by denys-hrulov — APPROVED** [review]

**Pull request merged**